### PR TITLE
fix(storage): fix GC watermark id

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Fix GC watermark after shared buffer refactor https://github.com/singularity-data/risingwave/pull/4773

- Previously we get synced epoch from `LocalVersion::shared_buffer`
- Now we should get synced epoch from `LocalVersion::sync_uncommitted_data`

## Checklist

- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/storage/src/hummock/local_version.rs [hunk 1]
```diff
@@ -322,9 +322,15 @@ impl LocalVersion {
...
 
     pub fn clear_shared_buffer(&mut self) -> Vec<HummockEpoch> {
-        let cleaned_epochs = self.shared_buffer.keys().cloned().collect_vec();
+        let mut cleaned_epoch = self.shared_buffer.keys().cloned().collect_vec();
+        for (epochs, _) in self.sync_uncommitted_data.iter() {
+            for epoch in epochs {
+                cleaned_epoch.push(*epoch);
+            }
+        }
+        self.sync_uncommitted_data.clear();
```
Comment (zwang28) [inline src/storage/src/hummock/local_version.rs:331]: Also removing sync_uncommitted_data here. @hzxa21 
What about replicated_batches?

Comment (hzxa21) [reply src/storage/src/hummock/local_version.rs:331]: We should also clear replicated_batches here.

Comment (zwang28) [reply src/storage/src/hummock/local_version.rs:331]: fixed

**Review by hzxa21 — APPROVED** [review]
Good catch! Thanks for the PR.

**Pull request merged**